### PR TITLE
[backend] Fix filters generation (#5370)

### LINKED ISSUE #5325: Can't do some api filters combinations in background tasks
Background tasks on relationships are not working for some api filters combinations.

Go to Data > Relationships. Select some filters. Select all the filtered relationships via the checkboxes. Do a background task (ex: marking definition update). The background task targets no entities with some filters combinations (see exemple below).

WORKING :
- a single filter
image
- a combination of classical filters
image
- api filters only on 'to' or only on 'from'
image

NOT WORKING:
- an api filter + a classical filter
image
- api filter(s) on 'from' and api filter(s) on 'to'
image

### LINKED ISSUE #5370: Error trying to export with specific API, fromId, toId, fromType etc.. - Data > Relationship
## Description

When selecting specials filters (fromId, toId, fromType, toType) and trying to export we got an error : "incorrect filter keys not existing in any schema definition"

image

## Reproducible Steps

Steps to create the smallest reproducible scenario:

1. Go to Data > Relationships page
2. Select some relationships
3. Click on top-right button "Open export panel"
4. Choose csv export and click "Create"

## Expected Output

Should export with the specials filters

## Actual Output

Got an error

### PR DESCRIPTION
**Pull request by richard-julien**
See #5370

### CODE REVIEW

**Comment by Archidoit** [comment]
The fix is working well
But errors remaining : 
- OR between an api filter and a classical filter is understand as an 'and'
Exemple: in the case below, only the relationship with (source_entity=menuPass AND relationship_type= targets) are exported
image

**Comment by Archidoit** [comment]
> The fix is working well But errors remaining :
> 
> * OR between an api filter and a classical filter is understand as an 'and'
>   Exemple: in the case below, only the relationship with (source_entity=menuPass AND relationship_type= targets) are exported
>   image

--> fixed

**Comment by Archidoit** [comment]
Background task on Data>Relationship with 1 filter on a 'to' and 1 filter on a 'from' : only 1 of the 2 filters in taken into account

Exemple , do a background task on the relationships with the filter : 
image
--> all the relationships with the given source entity are targeted (independently from their target type)

**Comment by Archidoit** [comment]
Background task on relationships with a classical filter and an api filter: working 👍 
exemple: 
image

**Comment by Archidoit** [comment]
Export of Data>Relationship: some complex combinations not working.
In the exemple below, all the relationships respecting the filter: 
[ ( target_type = attackPattern **OR** source_entity=menu pass ) OR relationship_type =uses ] are exported
image

File: opencti-platform/opencti-graphql/src/database/engine.js [hunk 1]
```diff
@@ -1965,43 +1977,76 @@ const completeSpecialFilterKeys = async (context, user, inputFilters) => {
...
+        const nested = [
+          { key: 'types', operator: filter.operator, values: filter.values },
+          { key: 'role', operator: 'wildcard', values: [`*_${side}`] }
+        ];
+        finalFilters.push({ key: 'connections', nested });
       }
-      if (newFilterGroup) {
-        finalFilterGroups.push(newFilterGroup);
+      if (arrayKeys.includes(INSTANCE_FILTER_TARGET_TYPES)) {
+        const nested = [{ key: 'types', operator: filter.operator, values: filter.values }];
```
Comment (Archidoit) [inline opencti-platform/opencti-graphql/src/database/engine.js:2004]: in buildRelationsFilter, if we also have a filter on elementId, we also add in the nested : { key: 'internal_id', values: optsElementIds, operator: FilterOperator.NotEq }

File: opencti-platform/opencti-graphql/src/database/engine.js [hunk 2]
```diff
@@ -1965,43 +1977,76 @@ const completeSpecialFilterKeys = async (context, user, inputFilters) => {
...
-      const { newFilter, newFilterGroup } = await adaptFilterToSourceReliabilityFilterKey(context, user, filter);
-      if (newFilter) {
-        finalFilters.push(newFilter);
+      if (arrayKeys.includes(INSTANCE_FILTER)) {
+        const nested = [{ key: 'internal_id', operator: filter.operator, values: filter.values }];
+        finalFilters.push({ key: 'connections', nested });
+      }
+      if (arrayKeys.includes(RELATION_FROM_FILTER) || arrayKeys.includes(RELATION_TO_FILTER)
+          || arrayKeys.includes(RELATION_TO_SIGHTING_FILTER)) {
+        const side = arrayKeys.includes(RELATION_FROM_FILTER) ? 'from' : 'to';
```
Comment (Archidoit) [inline opencti-platform/opencti-graphql/src/database/engine.js:2028]: if there is both a filter on fromId and on toId, only the from will be taken into account

File: opencti-platform/opencti-graphql/src/database/engine.js [hunk 3]
```diff
@@ -1965,43 +1977,76 @@ const completeSpecialFilterKeys = async (context, user, inputFilters) => {
...
+        const nested = [
+          { key: 'internal_id', operator: filter.operator, values: filter.values },
+          { key: 'role', operator: 'wildcard', values: [`*_${side}`] }
+        ];
+        finalFilters.push({ key: 'connections', nested });
+      }
+      if (arrayKeys.includes(RELATION_FROM_TYPES_FILTER) || arrayKeys.includes(RELATION_TO_TYPES_FILTER)) {
+        const side = arrayKeys.includes(RELATION_FROM_TYPES_FILTER) ? 'from' : 'to';
```
Comment (Archidoit) [inline opencti-platform/opencti-graphql/src/database/engine.js:2036]: same here for fromTypes and toTypes

**Review by Archidoit — APPROVED** [review]

File: opencti-platform/opencti-graphql/src/database/engine.js [hunk 4]
```diff
@@ -1878,43 +1892,83 @@ const completeSpecialFilterKeys = async (context, user, inputFilters) => {
...
-    if (arrayKeys.includes(TYPE_FILTER)) {
-      // in case we want to filter by entity_type
-      // we need to add parent_types checking (in case the given value in type is an abstract type)
-      const { newFilter, newFilterGroup } = adaptFilterToEntityTypeFilterKey(filter);
-      if (newFilter) {
-        finalFilters.push(newFilter);
+    if (arrayKeys.some((fiterKey) => complexConversionFilterKeys.includes(fiterKey))) {
+      if (arrayKeys.length > 1) {
+        throw Error(`A filter with these multiple keys is not supported: ${arrayKeys}`);
+      }
```
Comment (jpkha) [inline opencti-platform/opencti-graphql/src/database/engine.js]: I'm a bit confused about the necessity of this condition. Above, we verify whether we have one or more complexFilterKeys. If there are more than one, we throw an error.

File: opencti-platform/opencti-front/src/private/components/common/stix_core_relationships/views/EntityStixCoreRelationshipsEntitiesView.tsx [hunk 5]
```diff
@@ -122,23 +123,31 @@ EntityStixCoreRelationshipsEntitiesViewProps
...
+  // const backgroundTaskFilters = {
+  //     ...filters,
+  //     entity_type:
+  //       selectedTypes?.length > 0
+  //         ? selectedTypes.map((n) => ({ id: n, value: n }))
+  //         : [{ id: 'Stix-Core-Object', value: 'Stix-Core-Object' }],
+  //     [`rel_${selectedRelationshipTypes.at(0)}.*`]: [
+  //       { id: entityId, value: entityId },
+  //     ],
+  //   };
```
Comment (jpkha) [inline opencti-platform/opencti-front/src/private/components/common/stix_core_relationships/views/EntityStixCoreRelationshipsEntitiesView.tsx]: Is there any reason to keep it as a comment?

**Review by jpkha — APPROVED** [review]

**Comment by Archidoit** [comment]
RQ: 'regardingOf' est actuellement présent uniquement dans Data>Entities et dans Incidents.

**Comment by Archidoit** [comment]
Error at some entities creation:
- Fill an incident creation form
- Click on 'create'
- The creation form is grey
- You should reload the page (either: error in every screen)
- Then the incident appears created

Same for other entities: report, observed data, incident response, groupings etc

image

--> SOLVED

**Comment by Archidoit** [comment]
In the Observable tab of an entity, we also have non-observables: 
image

**Comment by Archidoit** [comment]
In an observable 'analysis' tab, we have all the containers (whereas we should only have the ones of the observable): 
image

**Pull request merged**